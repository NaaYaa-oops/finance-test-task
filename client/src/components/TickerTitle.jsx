import React, {memo} from 'react'
import {ListItemIcon, ListItemText} from '@mui/material'
import MonetizationOnIcon from '@mui/icons-material/MonetizationOn'
import {tickerStyles} from '../styles/tickerStyles'


export const TickerTitle = memo(({ticker}) => {
    return (
        <>
            <ListItemIcon>
                <MonetizationOnIcon/>
            </ListItemIcon>
            <ListItemText sx={tickerStyles.listItemText}>
                {ticker}
            </ListItemText>
        </>
    )
})

