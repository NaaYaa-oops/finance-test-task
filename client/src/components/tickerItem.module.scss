.stocks_up {
  color: #23a728;
  transition: all 1s ease-in-out;
}

.stocks_down {
  color: #FF5851;
  transition: all 1s ease-in-out;

}