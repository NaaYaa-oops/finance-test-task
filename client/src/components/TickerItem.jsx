import React, {memo} from 'react'
import CountUp from 'react-countup'
import {TickerTitle} from './TickerTitle'
import {IconButton, ListItem, ListItemText} from '@mui/material'
import DeleteIcon from '@mui/icons-material/Delete'
import {
    calculateChangedWorth,
    calculateDividendYield,
    dateToLocaleDate
} from '../helpers'
import cn from 'classnames'
import classes from './tickerItem.module.scss'
import {useActions} from '../hooks/useActions'


const TickerItem = (props) => {
    const {ticker, price, dividend, yield: stockYield, change, change_percent, last_trade_time} = props
    const stocks = cn({
        [classes.stocks_up]: stockYield >= 1,
        [classes.stocks_down]: stockYield < 1
    })
    const {deleteTickerByName} = useActions()
    const deleteHandler = event => {
        deleteTickerByName({ticker})
    }
    return (
        <ListItem>
            <TickerTitle ticker={ticker}/>
            <ListItemText>
                <CountUp
                    start={0}
                    end={price}
                    duration={1}
                    separator=" "
                    decimals={2}
                    decimal=","
                    prefix="$ "
                />
            </ListItemText>
            <ListItemText>
                <div className={stocks}>
                    {calculateChangedWorth(price, change, change_percent)}
                </div>
            </ListItemText>
            <ListItemText>
                <div className={stocks}>{calculateDividendYield(price,
                    dividend)}</div>
            </ListItemText>
            <ListItemText>
                {dateToLocaleDate(last_trade_time)}
            </ListItemText>
            <IconButton edge="end" aria-label="delete" onClick={deleteHandler}>
                <DeleteIcon/>
            </IconButton>
        </ListItem>
    )
}

export default memo(TickerItem)