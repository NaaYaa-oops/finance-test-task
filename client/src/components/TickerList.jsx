import React from 'react'
import {List} from '@mui/material'
import {useSelector} from 'react-redux'
import TickerItem from './TickerItem'


const TickerList = () => {
    const {tickers} = useSelector(state => state.ticker)

    return (
        <>
            <List>
                {tickers.map(
                    ticker => <TickerItem key={ticker.ticker}{...ticker}/>)}
            </List>
        </>
    )
}

export default TickerList