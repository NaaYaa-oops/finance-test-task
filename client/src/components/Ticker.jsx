import React from 'react'
import {TickerList} from './TickerList'
import {useSockets} from '../hooks/useSockets'


export const Ticker = () => {
    useSockets()

    return (
        <TickerList/>
    )
}

