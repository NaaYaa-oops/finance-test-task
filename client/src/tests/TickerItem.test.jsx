import {screen} from '@testing-library/react'
import {render} from './test-utils'
import TickerItem from '../components/TickerItem'
import {TickerTitle} from '../components/TickerTitle'


const state = {
    'ticker': 'AAPL',
    'exchange': 'NASDAQ',
    'price': 279.29,
    'change': 64.52,
    'change_percent': 0.84,
    'dividend': 0.56,
    'yield': 1.34,
    'last_trade_time': '2021-04-30T11:53:21.000Z'
}

describe('Ticker item component', () => {
    test('Ticker item renders', () => {
        render(<TickerItem/>)
        expect(screen.getByRole('listitem')).toBeInTheDocument()
    })
    test('Ticker item renders with value', () => {
        render(<TickerItem {...state}/>)
        expect(screen.getByText('AAPL')).toBeInTheDocument()
    })
    test('Ticker item snapshot', () => {
        const tickerItem = render(<TickerItem {...state}/>)
        expect(tickerItem).toMatchSnapshot()
    })
    test('Ticker item -> title renders without values', () => {
        render(<TickerTitle/>)
        expect(screen.queryByText((content, element) => {
            return element.tagName.toLowerCase() === 'span' && content.startsWith(
                'APPL')
        })).toBeNull()
    })
})