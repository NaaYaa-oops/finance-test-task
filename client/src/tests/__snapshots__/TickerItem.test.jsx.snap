// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ticker item component Ticker item snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="MonetizationOnIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1.41 16.09V20h-2.67v-1.93c-1.71-.36-3.16-1.46-3.27-3.4h1.96c.1 1.05.82 1.87 2.65 1.87 1.96 0 2.4-.98 2.4-1.59 0-.83-.44-1.61-2.67-2.14-2.48-.6-4.18-1.62-4.18-3.67 0-1.72 1.39-2.84 3.11-3.21V4h2.67v1.95c1.86.45 2.79 1.86 2.85 3.39H14.3c-.05-1.11-.64-1.87-2.22-1.87-1.5 0-2.4.68-2.4 1.64 0 .84.65 1.39 2.67 1.91s4.18 1.39 4.18 3.91c-.01 1.83-1.38 2.83-3.12 3.16z"
            />
          </svg>
        </div>
        <div
          class="MuiListItemText-root css-fgjcaj-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            AAPL
          </span>
        </div>
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            <span>
              $ 0,00
            </span>
          </span>
        </div>
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            <div
              class="stocks_up"
            >
              3.636$
            </div>
          </span>
        </div>
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            <div
              class="stocks_up"
            >
              0.20%
            </div>
          </span>
        </div>
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            4/30/2021
          </span>
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MonetizationOnIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1.41 16.09V20h-2.67v-1.93c-1.71-.36-3.16-1.46-3.27-3.4h1.96c.1 1.05.82 1.87 2.65 1.87 1.96 0 2.4-.98 2.4-1.59 0-.83-.44-1.61-2.67-2.14-2.48-.6-4.18-1.62-4.18-3.67 0-1.72 1.39-2.84 3.11-3.21V4h2.67v1.95c1.86.45 2.79 1.86 2.85 3.39H14.3c-.05-1.11-.64-1.87-2.22-1.87-1.5 0-2.4.68-2.4 1.64 0 .84.65 1.39 2.67 1.91s4.18 1.39 4.18 3.91c-.01 1.83-1.38 2.83-3.12 3.16z"
          />
        </svg>
      </div>
      <div
        class="MuiListItemText-root css-fgjcaj-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          AAPL
        </span>
      </div>
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          <span>
            $ 0,00
          </span>
        </span>
      </div>
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          <div
            class="stocks_up"
          >
            3.636$
          </div>
        </span>
      </div>
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          <div
            class="stocks_up"
          >
            0.20%
          </div>
        </span>
      </div>
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          4/30/2021
        </span>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
